### PR TITLE
Add test for casting after a clean or soft cast

### PR DESCRIPTION
**Pull request by smowton**
This is a test for https://github.com/diffblue/cbmc/pull/3822/commits/0a3b99ba59267c3bf681a24ce0549394ca550c47, one of the type consistency fixes committed with https://github.com/diffblue/cbmc/pull/3822

These will get easier to express as and when `expr2java` prints generic type parameters / qualifiers, but I won't take that task on here, particularly because that means checking all consumers of pretty-printed output.

### CODE REVIEW

File: jbmc/regression/jbmc/generic_base_type/base.java [hunk 1]
```diff
@@ -0,0 +1,21 @@
+
+public class base<T> { }
```
Comment (peterschrammel) [inline jbmc/regression/jbmc/generic_base_type/base.java]: Please capitalize class identifiers

Comment (peterschrammel) [reply jbmc/regression/jbmc/generic_base_type/base.java]: base.class (Base.class) file doesn't seem to be committed ?

**Review by peterschrammel — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 5136721987c280f8339bf46c02d8d49d5fa4eb80).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101129063
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 265b96b57aaaf396e9b80a20b2d3885cf8e393a7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101131620

**Review by allredj — COMMENTED** [review]
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: f6bbc9a3ab9df25511f2b86d71871cd8a2f750cc).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/101130945
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

**Pull request merged**